src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,222 @@
+import org.embulk.config.ConfigDefault;
+import org.embulk.config.ConfigDiff;
+import org.embulk.config.ConfigException;
+import org.embulk.config.TaskReport;
+import org.embulk.input.marketo.MarketoService;
+import org.embulk.input.marketo.MarketoUtils;
+import org.embulk.spi.Schema;
+import org.embulk.spi.type.Types;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+
+public class ProgramInputPlugin extends MarketoBaseInputPluginDelegate<ProgramInputPlugin.PluginTask>
+{
+    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormat.forPattern(MarketoUtils.MARKETO_DATE_SIMPLE_DATE_FORMAT);
+
+    public interface PluginTask extends MarketoBaseInputPluginDelegate.PluginTask
+    {
+        @Config("query_by")
+        @ConfigDefault("null")
+        Optional<QueryBy> getQueryBy();
+
+        @Config("tag_type")
+        @ConfigDefault("null")
+        Optional<String> getTagType();
+
+        @Config("tag_value")
+        @ConfigDefault("null")
+        Optional<String> getTagVallue();
+
+        @Config("earliest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getEarliestUpdatedAt();
+
+        @Config("latest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getLatestUpdatedAt();
+
+        @Config("filter_type")
+        @ConfigDefault("null")
+        Optional<String> getFilterType();
+
+        @Config("filter_values")
+        @ConfigDefault("null")
+        Optional<List<String>> getFilterValues();
+
+        @Config("report_duration")
+        @ConfigDefault("null")
+        Optional<Long> getReportDuration();
+
+        void setLatestUpdatedAt(Optional<Date> latestUpdatedAt);
+    }
+
+    public ProgramInputPlugin()
+    {
+    }
+
+    @Override
+    public void validateInputTask(PluginTask task)
+    {
+        super.validateInputTask(task);
+        // validate if query_by is selected
+        if (task.getQueryBy().isPresent()) {
+            switch(task.getQueryBy().get()) {
+            case TAG_TYPE:
+                //make sure tag type and tag value are not empty
+                if (!task.getTagType().isPresent() || !task.getTagVallue().isPresent()) {
+                    throw new ConfigException("tag_type and tag_value are required when query by Tag Type");
+                }
+                break;
+            case DATE_RANGE:
+                // make sure earliest_updated_at is not empty
+                if (!task.getEarliestUpdatedAt().isPresent()) {
+                    throw new ConfigException("`earliest_updated_at` is required when query by Date Range");
+                }
+
+                DateTime earliest = new DateTime(task.getEarliestUpdatedAt().get());
+                // make sure latest_updated_at is not empty
+                if (!task.getLatestUpdatedAt().isPresent()) {
```

Comment:
I guess it should combinative condition here, something like: `if (!task.getLatestUpdatedAt.isPresent || !task.getReportDuration().isPresent())`

Reply:
The latest update should always present even with incremental import because it's used to calculate the report duration

Reply:
what if below configuration:

```
incremental: true
earliest_updated_at: xxx
report_duration: xxx
```

It's the acceptable case but with the current implementation, exception will be thrown